File: lib/qni/export/circle_notation_png.rb
# frozen_string_literal: true

require 'fileutils'
require 'json'
require 'open3'
require 'tmpdir'
require_relative '../simulator'

module Qni
  module Export
    # Invokes the Python helper that renders computational-basis circle-notation PNG files.
    class CircleNotationPng
      SETUP_MESSAGE = 'circle-notation rendering requires matplotlib and Pillow; run scripts/setup_symbolic_python.sh'
      HELPER_RELATIVE_PATH = '../../../libexec/qni_circle_notation_render.py'
      REPO_RUNTIME_RELATIVE_PATH = '../../../.python-symbolic/bin/python'
      HELPER_ENV = {
        'MPLCONFIGDIR' => File.join(Dir.tmpdir, 'qni-cli-matplotlib'),
        'UV_CACHE_DIR' => File.join(Dir.tmpdir, 'qni-cli-uv-cache')
      }.freeze

      def initialize(state_vector:, output_path:, theme:)
        @state_vector = state_vector
        @output_path = output_path
        @theme = theme
      end

      def export
        FileUtils.mkdir_p(File.dirname(output_path))
        result = render_with_helper_commands
        return if result == :file_rendered
        return result if result

        raise Simulator::Error, SETUP_MESSAGE
      end

      private

      attr_reader :output_path, :state_vector, :theme

      def helper_commands
        [
          [repo_runtime_path, helper_path],
          ['python3', helper_path]
        ]
      end

      def helper_path
        File.expand_path(HELPER_RELATIVE_PATH, __dir__)
      end

      def repo_runtime_path
        File.expand_path(REPO_RUNTIME_RELATIVE_PATH, __dir__)
      end

      def payload
        state_vector.export_payload.merge(
          'output_path' => output_path,
          'theme' => theme.to_s
        )
      end

      def run_with_helper(command)
        _stdout, stderr, status = Open3.capture3(
          HELPER_ENV,
          *command,
          stdin_data: JSON.generate(payload)
        )
        return :file_rendered if status.success?

        handle_failed_helper(stderr, status)
      rescue Errno::ENOENT
        :retry_with_next_command
      end

      def render_with_helper_commands
        helper_commands.each do |command|
          result = run_with_helper(command)
          return result unless result == :retry_with_next_command
        end

        nil
      end

      def handle_failed_helper(stderr, status)
        renderer_class = self.class
        return :retry_with_next_command if renderer_class.retryable_with_next_command?(stderr)

        raise Simulator::Error, renderer_class.helper_error_message(stderr, status)
      end

      class << self
        def helper_error_message(stderr, status)
          message = stderr.to_s.strip
          return "circle-notation renderer failed with exit status #{status.exitstatus}" if message.empty?

          message
        end

        def retryable_with_next_command?(stderr)
          message = stderr.to_s
          message.include?("No module named 'matplotlib'") ||
            message.include?("No module named 'PIL'")
        end
      end
    end
  end
end

File: src/export/circle_notation_png.ts
import { spawnSync } from 'node:child_process';
import { mkdirSync } from 'node:fs';
import { tmpdir } from 'node:os';
import path from 'node:path';

import type { StateVectorExportPayload } from '../simulator';
import type { ExportTheme } from './qcircuit_latex';

export class CircleNotationPngError extends Error {}

export interface CircleNotationPngOptions {
  readonly env?: NodeJS.ProcessEnv;
  readonly outputPath: string;
  readonly projectRoot: string;
  readonly stateVector: StateVectorExportPayload;
  readonly theme: ExportTheme;
}

interface CircleNotationRenderPayload extends StateVectorExportPayload {
  readonly output_path: string;
  readonly theme: ExportTheme;
}

interface HelperCommand {
  readonly args: readonly string[];
  readonly command: string;
}

interface HelperResult {
  readonly error?: NodeJS.ErrnoException;
  readonly status: number | null;
  readonly stderr: string;
}

const SETUP_MESSAGE = 'circle-notation rendering requires matplotlib and Pillow; run scripts/setup_symbolic_python.sh';

export class CircleNotationPng {
  private readonly env: NodeJS.ProcessEnv;
  private readonly outputPath: string;
  private readonly projectRoot: string;
  private readonly stateVector: StateVectorExportPayload;
  private readonly theme: ExportTheme;

  constructor(options: CircleNotationPngOptions) {
    this.env = options.env ?? {};
    this.outputPath = options.outputPath;
    this.projectRoot = options.projectRoot;
    this.stateVector = options.stateVector;
    this.theme = options.theme;
  }

  export(): void {
    mkdirSync(path.dirname(this.outputPath), { recursive: true });

    for (const command of this.helperCommands()) {
      const result = this.runWithHelper(command);

      if (result === 'file-rendered') {
        return;
      }
    }

    throw new CircleNotationPngError(SETUP_MESSAGE);
  }

  private helperCommands(): HelperCommand[] {
    const helperPath = path.join(this.projectRoot, 'libexec', 'qni_circle_notation_render.py');

    return [
      {
        args: [helperPath],
        command: path.join(this.projectRoot, '.python-symbolic', 'bin', 'python')
      },
      {
        args: [helperPath],
        command: 'python3'
      }
    ];
  }

  private runWithHelper(command: HelperCommand): 'file-rendered' | 'retry-with-next-command' {
    const result = this.captureHelper(command);

    if (result.error?.code === 'ENOENT') {
      return 'retry-with-next-command';
    }

    if (result.error) {
      throw new CircleNotationPngError(result.error.message);
    }

    if (result.status === 0) {
      return 'file-rendered';
    }

    if (retryableWithNextCommand(result.stderr)) {
      return 'retry-with-next-command';
    }

    throw new CircleNotationPngError(helperErrorMessage(result.stderr, result.status));
  }

  private captureHelper(command: HelperCommand): HelperResult {
    const result = spawnSync(command.command, [...command.args], {
      encoding: 'utf8',
      env: {
        ...process.env,
        ...this.env,
        MPLCONFIGDIR: path.join(tmpdir(), 'qni-cli-matplotlib'),
        UV_CACHE_DIR: path.join(tmpdir(), 'qni-cli-uv-cache')
      },
      input: JSON.stringify(this.payload()),
      stdio: ['pipe', 'pipe', 'pipe']
    });

    return {
      error: result.error as NodeJS.ErrnoException | undefined,
      status: result.status,
      stderr: result.stderr ?? ''
    };
  }

  private payload(): CircleNotationRenderPayload {
    return {
      ...this.stateVector,
      output_path: this.outputPath,
      theme: this.theme
    };
  }
}

function retryableWithNextCommand(stderr: string): boolean {
  return stderr.includes("No module named 'matplotlib'") || stderr.includes("No module named 'PIL'");
}

function helperErrorMessage(stderr: string, status: number | null): string {
  const message = stderr.trim();

  return message.length === 0 ? `circle-notation renderer failed with exit status ${status ?? ''}` : message;
}
